src/include/migraphx/op/prefix_scan_op.hpp

```diff
@@ -0,0 +1,111 @@
+    {
+        return pack(f(self.axis, "axis"), 
+                    f(self.exclusive, "exclusive"), 
+                    f(self.reverse, "reverse"));
+    }
+
+    value attributes() const
+    {
+        value normalize;
+        normalize["axis"] = value::array{normalize_attribute::include_min};
+        return {{"normalize_axes", normalize}};
+    }
+
+    shape normalize_compute_shape(std::vector<shape> inputs) const
+    {
+        check_shapes{inputs, *this}.has(1);
+        auto s          = inputs.at(0);
+
+        return {s.type(), s.lens()};
+    }
+
+    void prefix_scan(argument& arg, argument& res) const
+    {
+        auto& self = static_cast<const Derived&>(*this);
+        if (reverse) {
+            visit_all(arg, res)([&](auto input, auto output) { 
+                shape_for_each_reverse(output.get_shape(), [&](const auto& idx) {
+                    if (idx[axis] == arg.get_shape().lens()[axis] - 1) {
+                        if (exclusive) 
+                            output(idx.begin(), idx.end()) = 0;
+                        else
+                            output(idx.begin(), idx.end()) = input(idx.begin(), idx.end());
+                    }
+                    else {
+                        auto prefix_idx = idx;
+                        prefix_idx[axis] += 1;
+                        if (exclusive) 
+                            output(idx.begin(), idx.end()) = self.op()(input(prefix_idx.begin(), prefix_idx.end()), output(prefix_idx.begin(), prefix_idx.end()));
+                        else
+                            output(idx.begin(), idx.end()) = self.op()(input(idx.begin(), idx.end()), output(prefix_idx.begin(), prefix_idx.end()));
+                    }
+                });
+            });
+        }
+        else {
+            visit_all(arg, res)([&](auto input, auto output) { 
+                shape_for_each(output.get_shape(), [&](const auto& idx) {
+                    if (idx[axis] == 0) {
+                        if (exclusive) 
+                            output(idx.begin(), idx.end()) = 0;
+                        else
+                            output(idx.begin(), idx.end()) = input(idx.begin(), idx.end());
+                    }
+                    else {
+                        auto prefix_idx = idx;
+                        prefix_idx[axis] -= 1;
+                        if (exclusive) 
+                            output(idx.begin(), idx.end()) = self.op()(input(prefix_idx.begin(), prefix_idx.end()), output(prefix_idx.begin(), prefix_idx.end()));
+                        else
+                            output(idx.begin(), idx.end()) = self.op()(input(idx.begin(), idx.end()), output(prefix_idx.begin(), prefix_idx.end()));
+                    }
+                });
+            });
+        }
+    }
+    
+    argument compute(const shape& output_shape, std::vector<argument> args) const
+    {
+        argument result{output_shape};
+        this->prefix_scan(args[0], result);
```

Comment:
This could be done in-place:

```cpp
argument result = args[0];
this->prefix_scan(result);
```

Reply:
Actually, this could be done in parallel by slicing the shape and then doing `partial_sum` in the inner loop since we are only doing this across one axis. A rough idea of how this can be done:

```cpp
argument result = args[0];
auto s = result.get_shape();
auto lens = s.lens();
auto size = lens[axis];
lens[axis] = 1;
auto slice = shape{s.type(), lens, s.strides()};
result.visit([&](auto output) {
    par_for(slice.elements(), [&](auto i) {
        // Get pointer to first element for prefix_sum
        auto* start = output.data()+slice.index(i);
        std::partial_sum(start, start+size, start, self.op());
    });
});
```

This will need to be adjusted for an exclusive scan.

Reply:
also the reverse can be done with `std::partial_sum(std::make_reverse_iterator(start), std::make_reverse_iterator(start+size), start, self.op())`

Reply:
> This could be done in-place:
> 
> ```c++
> argument result = args[0];
> this->prefix_scan(result);
> ```

Exclusive mode either requires two copies or either a preliminary or follow-up pass to shift the elements.

Reply:
> Actually, this could be done in parallel by slicing the shape and then doing `std::partial_sum` in the inner loop since we are only doing this across one axis. A rough idea of how this can be done:
> 
> ```c++
> argument result = args[0];
> auto s = result.get_shape();
> auto lens = s.lens();
> auto size = lens[axis];
> lens[axis] = 1;
> auto slice = shape{s.type(), lens, s.strides()};
> result.visit([&](auto output) {
>     par_for(slice.elements(), [&](auto i) {
>         // Get pointer to first element for prefix_sum
>         auto* start = output.data()+slice.index(i);
>         std::partial_sum(start, start+size, start, self.op());
>     });
> });
> ```
> 
> This will need to be adjusted for an exclusive scan.

The issue that seems to be happening with using partial_sum is that the elements are only contiguous when `axis = dimensions - 1`. 

Also, I think we can parallelize along all axes other than the one being summed along. So for 3-D and axis=0, both inner loops can be parallelized; 3-D and axis=1, the outer and inner loops can be parallelized; or 3-D and axis=2, the two outer loops can be parallelized. I have an idea for how to do this, but I think it would require having a maximum number of dimensions and padding tensors smaller than that with size 1 dimensions. Can we assume 3 is the max?

Reply:
> The issue that seems to be happening with using partial_sum is that the elements are only contiguous when axis = dimensions - 1.

That's right, it only works if the strides are 1 for the dimension. We need to slice it into another tensor_view:

```cpp
argument result = args[0];
auto s = result.get_shape();
auto slice = shape{s.type(), {s.lens()[axis]}, {s.strides()[axis]}};
auto lens = s.lens();
lens[axis] = 1;
auto batch = shape{s.type(), lens, s.strides()};
result.visit([&](auto output) {
    using type = decltype(output);
    par_for(batch.elements(), [&](auto i) {
        auto* start = output.data()+batch.index(i);
        type x{slice, start};
        std::partial_sum(x.begin(), x.end(), x.begin(), self.op());
    });
});
```

> Can we assume 3 is the max?

No we can't.